pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentReplicator.java

```diff
@@ -364,16 +378,31 @@ public void readEntriesComplete(List<Entry> entries, Object ctx) {
 
                 headersAndPayload.retain();
 
-                getSchemaInfo(msg).thenAccept(schemaInfo -> {
-                    msg.setSchemaInfoForReplicator(schemaInfo);
+                CompletableFuture<SchemaInfo> schemaFuture = getSchemaInfo(msg);
+                if (!schemaFuture.isDone() || schemaFuture.isCompletedExceptionally()) {
+                    // Mark the replicator is fetching the schema for now and rewind the cursor
+                    // and trigger the next read after complete the schema fetching.
+                    fetchSchemaInProgress = true;
+                    skipRemainingMessages = true;
```

Comment:
Do we need to release the first entry here?

Reply:
Oh, nice catch!
Yes, we need to release the first entry.